File: api/app/services/bullets.py
import re
from pathlib import Path
from functools import lru_cache
from app.models.schemas import BulletFeedback


@lru_cache(maxsize=1)
def _load_verbs() -> frozenset[str]:
    verbs_path = Path(__file__).parent.parent.parent / "data" / "action_verbs.txt"
    return frozenset(
        line.strip().lower()
        for line in verbs_path.read_text().splitlines()
        if line.strip()
    )


# Patterns for passive / weak bullet openings
_PASSIVE_PATTERNS = [
    re.compile(r"^was responsible for", re.I),
    re.compile(r"^responsible for", re.I),
    re.compile(r"^helped (with|to)", re.I),
    re.compile(r"^assisted (with|in)", re.I),
    re.compile(r"^worked on", re.I),
    re.compile(r"^participated in", re.I),
    re.compile(r"^involved in", re.I),
    re.compile(r"^duties included", re.I),
    re.compile(r"^tasked with", re.I),
]

# Numbers, percentages, dollar amounts, multipliers
_METRIC_RE = re.compile(
    r"\b\d[\d,.]*\s*(%|x|×|times|k|m|b|million|billion|\$|dollars?|usd)\b|\$[\d,.]+|\b\d{2,}\b",
    re.I,
)


def audit_bullets(experience_text: str) -> list[BulletFeedback]:
    verbs = _load_verbs()

    lines = [l.strip() for l in experience_text.splitlines() if l.strip()]
    # Accept lines that start with a bullet marker or are long enough to be a bullet
    bullet_lines = [
        l for l in lines
        if re.match(r"^[•\-\*\u2022\u2013]", l) or len(l.split()) >= 5
    ]

    if not bullet_lines:
        return []

    feedback: list[BulletFeedback] = []
    for raw in bullet_lines:
        text = re.sub(r"^[•\-\*\u2022\u2013]\s*", "", raw)
        issues: list[str] = []

        # 1. Strong action verb
        first_word = re.sub(r"[^a-z]", "", (text.split()[0] if text.split() else "").lower())
        if first_word and first_word not in verbs:
            issues.append(f'Starts with "{first_word}" — consider a strong action verb')

        # 2. Measurable impact
        if not _METRIC_RE.search(text):
            issues.append("No measurable impact — add a number or percentage")

        # 3. Too brief
        if len(text.split()) < 8:
            issues.append("Too brief — expand with context or outcome")

        # 4. Passive framing
        for pattern in _PASSIVE_PATTERNS:
            if pattern.match(text):
                issues.append("Passive framing — lead with what you did, not your role")
                break

        feedback.append(BulletFeedback(text=text, passed=len(issues) == 0, issues=issues))

    return feedback

File: lib/analyzer/bullets.ts
import { readFileSync } from 'fs'
import path from 'path'
import type { BulletFeedback } from '@/types/analysis'

let _verbs: Set<string> | null = null

function loadVerbs(): Set<string> {
  if (_verbs) return _verbs
  const filePath = path.join(process.cwd(), 'data', 'action_verbs.txt')
  const raw = readFileSync(filePath, 'utf-8')
  _verbs = new Set(
    raw.split('\n').map(l => l.trim().toLowerCase()).filter(Boolean),
  )
  return _verbs
}

// Passive / weak openings that signal a bullet needs rewording
const PASSIVE_PATTERNS = [
  /^was responsible for/i,
  /^responsible for/i,
  /^helped (with|to)/i,
  /^assisted (with|in)/i,
  /^worked on/i,
  /^participated in/i,
  /^involved in/i,
  /^duties included/i,
  /^tasked with/i,
]

// Metrics: numbers, percentages, dollar amounts, multipliers
const METRIC_RE = /\b\d[\d,.]*\s*(%|x|×|times|k|m|b|million|billion|\$|dollars?|usd)\b|\$[\d,.]+|\b\d{2,}\b/i

export function auditBullets(experienceText: string): BulletFeedback[] {
  const verbs = loadVerbs()

  // Split on lines that start with a bullet marker or are non-empty after trimming
  const lines = experienceText
    .split('\n')
    .map(l => l.trim())
    .filter(l => l.length > 0)

  // Keep only lines that look like bullets (start with marker or are long enough)
  const bulletLines = lines.filter(l =>
    /^[•\-\*\u2022\u2013]/.test(l) || l.split(/\s+/).length >= 5,
  )

  if (bulletLines.length === 0) return []

  return bulletLines.map((raw): BulletFeedback => {
    const text = raw.replace(/^[•\-\*\u2022\u2013]\s*/, '')
    const issues: string[] = []

    // 1. Strong action verb check
    const firstWord = text.split(/\s+/)[0]?.toLowerCase().replace(/[^a-z]/g, '')
    if (firstWord && !verbs.has(firstWord)) {
      issues.push(`Starts with "${firstWord}" — consider a strong action verb`)
    }

    // 2. Metrics check
    if (!METRIC_RE.test(text)) {
      issues.push('No measurable impact — add a number or percentage')
    }

    // 3. Length check (fewer than 8 words is usually too vague)
    const wordCount = text.split(/\s+/).filter(Boolean).length
    if (wordCount < 8) {
      issues.push('Too brief — expand with context or outcome')
    }

    // 4. Passive / weak framing
    const passiveMatch = PASSIVE_PATTERNS.find(p => p.test(text))
    if (passiveMatch) {
      issues.push('Passive framing — lead with what you did, not your role')
    }

    return { text, passed: issues.length === 0, issues }
  })
}
